### ISSUE TITLE
Enabling CONFIG_FP_SHARING on ARM causes usage fault

### ISSUE THREAD

**Issue by nashif** [zephyriot/zep-jira14#1341]
**_Reported by Florian Vaussard:_**

I am using an STM32F412 (Cortex-M4) based on the official STM32F4 port (tested with both version 1.6 and today's head). I enabled CPU_HAS_FPU in stm32f4 arch. Running with CONFIG_FLOAT works, but enabling CONFIG_FP_SHARING will causes a usage fault. One might think that this is related to missing FPU support in STM32F4, but I happened to test it several months ago (before the unified kernel) using my own STM32F4 port (as the official's one was not yet merged), and it worked just fine. Indeed, the FPU support is related to the Cortex-M4 core support and has nothing STM-specific. Thus I consider it to be a regression, even if it has never been officially enabled/supported.

The test application is really simple; just a dummy infinite loop with a k_sleep() inside the main thread. The fault occurs around the 2nd context switch, even if no floating point arithmetic is used at all. Here is a sample output of the console:

```
Plop 0.100000
Plop 0.200000
***** USAGE FAULT *****
  Executing thread ID (thread): 0x200007a0
  Faulting instruction address:  0x0
  Illegal use of the EPSR
Fatal fault in essential thread! Spinning...
```

I did a quick debug using JTAG, but nothing conclusive came out. PC apparently jumps to 0x0 after the "bx lr" at the end of __pendsv, but I do not know if gdb gives me accurate information as LR seems to equal 0xffffffed at this point, regardless of CONFIG_FP_SHARING, which left me wondering how this could work at all. I did not had time to debug further, but maybe someone will have an idea?

(Imported from Jira ZEP-1470)

**Comment by nashif** [comment]
**_by Benjamin Walsh:_**

_Florian Vaussard_  "as LR seems to equal 0xffffffed"

Is this a typo and should be 0xFFFFFFFD or is it really 0xFFFFFFED ? Because the first one is a valid return address and the second one is not. Just making sure before starting a wild goose chase.

**Comment by nashif** [comment]
**_by Florian Vaussard:_**

Yes it is really 0xFFFFFFED, but it is a legal value for Cortex-M4 when FPU is enabled (see http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.dui0553a/Babefdjc.html) and it is not the reason. I debugged a bit more, and I found that the stack frame is corrupted when FPU registers are saved.

In my simple program, the second execution of __pendsv happens when the idle task is scheduled out in order to schedule in the main thread. When entering the exception handler, the situation as follow:

```
(gdb) monitor regs
R0 = 00000000, R1 = E000ED04, R2 = 10000000, R3 = 00000400
R4 = 08002EF1, R5 = AAAAAAAA, R6 = AAAAAAAA, R7 = 20000828
R8 = AAAAAAAA, R9 = AAAAAAAA, R10= AAAAAAAA, R11= AAAAAAAA
R12= AAAAAAAA, R13= 20001038, MSP= 20001038, PSP= 200007C0
R14(LR) = FFFFFFED, R15(PC) = 08001FB0
XPSR 6100000E, APSR 60000000, EPSR 01000000, IPSR 0000000E

(gdb) x /16xw 0x200007C0
0x200007c0 <_idle_stack+136>:   0x00000000      0x00000020      0x00000000      0x00000000
0x200007d0 <_idle_stack+152>:   0xaaaaaaaa      0x08002f03      0x080023e4      0x41000000
0x200007e0 <_idle_stack+168>:   0xaaaaaaaa      0xaaaaaaaa      0xaaaaaaaa      0xaaaaaaaa
0x200007f0 <_idle_stack+184>:   0xaaaaaaaa      0xaaaaaaaa      0xaaaaaaaa      0xaaaaaaaa
```

Thus bottom of the idle task's stack is 0x200007C0, and the saved PC is stored at 200007D8 (= 0x080023e4 which corresponds to arch/arm/core/cpu_idle.S:140). But when saving the task's FPU registers, R0 equals 0x200007A0 (pointer to _kernel.current->arch.preempt_float), thus the call to

```
vstmia r0, {s16-s31}
```

will write 64 bytes and overwrite the memory range 0x200007A0->0x200007DF, corrupting the stack frame.

```
(gdb) x /16xw 0x200007C0
0x200007c0 <_idle_stack+136>:   0xaaaaaaaa      0xaaaaaaaa      0xaaaaaaaa      0xaaaaaaaa
0x200007d0 <_idle_stack+152>:   0x08002ef1      0x00000000      0x00000000      0x00000000
0x200007e0 <_idle_stack+168>:   0xaaaaaaaa      0xaaaaaaaa      0xaaaaaaaa      0xaaaaaaaa
0x200007f0 <_idle_stack+184>:   0xaaaaaaaa      0xaaaaaaaa      0xaaaaaaaa      0xaaaaaaaa
```

Everything continues, until the main thread is scheduled out and the idle task is scheduled in during the 3rd call to __pendsv. When performing the exc return, the corrupted stack is installed and PC will jump to *0x200007d8 = 0x00000000.

Thus the problem is the stack corruption when writing to the task's arch.preempt_float[] array. The struct k_thread is located too close from the bottom of the stack.

As you will probably ask, size of the stacks is set to default (256 for idle thread, 2048 for ISR), except for the main thread which is set to 2048 using prj.mdef.

**Comment by nashif** [comment]
**_by Florian Vaussard:_**

(replying to my-self...)

Increasing the idle stack size will fix the corruption. We could do something like this:

```
--- a/kernel/unified/Kconfig
+++ b/kernel/unified/Kconfig
@@ -134,7 +134,7 @@ config IDLE_STACK_SIZE
        int
        prompt "Size of stack for idle thread"
        default 256
-       default 320 if ARC
+       default 320 if ARC || ARM
        help
        Depending on the work that the idle task must do, most likely due to
        power management but possibly to other features like system event
```

**Comment by nashif** [comment]
**_by Benjamin Walsh:_**

_Florian Vaussard_  I did not know about the different return values for when FPU is enabled (0xFFFFFFEx). You learn something every day.

The stack space for the idle thread should probably be increased only if (ARM && FP_SHARING), no ?

**Comment by nashif** [comment]
**_by Florian Vaussard:_**

Do not worry, I also learned something ;)

I also thought about doing this. The problem with "default" in Kconfig is that FP_SHARING should be in your defconfig, otherwise I think that the value will not be updated in your .config if you select FP_SHARING through menuconfig afterwards. Ideally we should enforce the minimum acceptable value. Maybe using a "range" property?

I also saw an oddity. The struct _preempt_float is added in struct _thread_arch when CONFIG_FLOAT is set. But it will remain unused until CONFIG_FP_SHARING is set. Maybe it would be more logical to include it only when FP_SHARING is set, thus saving some space when only CONFIG_FLOAT is set?

**Comment by nashif** [comment]
**_by Anas Nashif:_**

ARM specific issue

**Comment by nashif** [comment]
**_by Mark Linkmeyer:_**

Correcting the priority field

**Comment by nashif** [comment]
**_by Mark Linkmeyer:_**

Hi _Maureen Helm_ , is this bug sufficiently described such that investigation into fixing it can begin when the assignee (you) gets to it?  If so, will you please move it to the To Do state so it's in the queue to get fixed?  If not, can you ask the reporter for clarifying information so it can become ready for investigation?  I'm asking because it's a medium priority bug and we need to get the medium priority bug count down to release 1.7.  Thanks!

**Comment by nashif** [comment]
**_by Maureen Helm:_**

Fix was merged into master ahead of v.1.8 release, and cherry-picked to 1.6 and 1.7 release branches.
